### ISSUE TITLE
Proposal: add an optional copy argument to np.reshape

### ISSUE THREAD

**Issue by shoyer** [numpy/numpy#9818]
Inspired by my discussion with @charris in #9810, I would like to propose adding `copy` argument to `np.reshape()` and `ndarray.reshape()`.

Proposed semantics:
- `copy=None`: copy only if necessary (default).
- `copy=True`: always copy.
- `copy=False`: never copy. Raise an error if the data cannot be reshaped without a copy.

This would allow users to ensure desired behavior for `reshape` and alleviate the primary use case for modifying `ndarray.shape`.

**Comment by seberg** [comment]
Always agreed with that, the only slight annoyance is that we need yet another C-API version probably.

**Comment by eric-wieser** [comment]
Yep, this seems like a very good idea to me too.

**Comment by mhvk** [comment]
Agreed too, the only reason I use assignments to shape is because in some cases I want to be absolutely sure there won't be a copy for a very large array. I can similarly see the use of being able to ensure a copy is in fact made.

**Comment by seberg** [comment]
@mhvk and the fun thing is, it does not really help you (except in the sense of an assert statement), because the shape assignment actually *does* the copy, then sees its a copy and raises an error.

**Comment by charris** [comment]
ISTM that reshape can also replace ravel.

**Comment by charris** [comment]
Note that we will need to make sure that this doesn't break current subclasses of ndarray that do not implement the new keyword.

**Comment by seberg** [comment]
True, about ravel, there is a problem with ravel, in that it makes sure that the output array is contiguous, which was a regression we once had and because someone (maybe scipy/some scikit) relied on it I think.

**Comment by shoyer** [comment]
On the C-API side, I suppose it would make sense to use enums here rather PyObject for the extra argument? e.g., `NPY_COPY_DEFAULT`, `NPY_COPY_TRUE`, `NPY_COPY_FALSE`?

**Comment by rgommers** [comment]
In gh-11884, @seberg proposed the same thing as this issue. With the extra motivation that users now modify `arr.shape` instead of using `reshape`.
